libs/collab-stream/src/model.rs

```diff
@@ -355,8 +367,266 @@ impl TryFrom<CollabUpdateEvent> for StreamBinary {
+  }
+}
+
+pub struct AwarenessStreamUpdate {
+  pub data: Vec<u8>, // AwarenessUpdate::encode_v1
+  pub sender: CollabOrigin,
+}
+
+impl AwarenessStreamUpdate {
+  /// Returns Redis stream key, that's storing entries mapped to/from [AwarenessStreamUpdate].
+  pub fn stream_key(workspace_id: &str, object_id: &str) -> String {
+    format!("af:{}:{}:awareness", workspace_id, object_id)
+  }
+}
+
+pub(crate) struct AwarenessStreamUpdateBatch {
+  pub updates: BTreeMap<MessageId, AwarenessStreamUpdate>,
+}
+impl FromRedisValue for AwarenessStreamUpdateBatch {
+  fn from_redis_value(v: &Value) -> RedisResult<Self> {
+    let sr: SRRows = SRRows::from_redis_value(v)?;
+    let mut updates = BTreeMap::new();
+    for stream in sr {
+      for (_stream_key, messages) in stream {
+        for message in messages {
+          for (message_id, fields) in message {
+            let message_id =
+              MessageId::try_from(message_id).map_err(|e| internal(e.to_string()))?;
+            let sender = match fields.get("sender") {
+              None => CollabOrigin::Empty,
+              Some(sender) => {
+                let raw_origin = String::from_redis_value(sender)?;
+                let collab_origin = collab_origin_from_str(&raw_origin)?;
+                collab_origin
+              },
+            };
+            let data_raw = fields
+              .get("data")
+              .ok_or_else(|| internal("expecting field `data`"))?;
+            let data: Vec<u8> = FromRedisValue::from_redis_value(data_raw)?;
+            updates.insert(message_id, AwarenessStreamUpdate { data, sender });
+          }
+        }
+      }
+    }
+
+    Ok(AwarenessStreamUpdateBatch { updates })
+  }
+}
+
+//FIXME: this should be `impl FromStr for CollabOrigin`
+fn collab_origin_from_str(value: &str) -> RedisResult<CollabOrigin> {
+  match value {
+    "" => Ok(CollabOrigin::Empty),
+    "server" => Ok(CollabOrigin::Server),
+    other => {
+      let mut split = other.split('|');
+      match (split.next(), split.next()) {
+        (Some(uid), Some(device_id)) | (Some(device_id), Some(uid))
+          if uid.starts_with("uid:") && device_id.starts_with("device_id:") =>
+        {
+          let uid = uid.trim_start_matches("uid:");
+          let device_id = device_id.trim_start_matches("device_id:").to_string();
+          let uid: i64 = uid
+            .parse()
+            .map_err(|err| internal(format!("failed to parse uid: {}", err)))?;
+          Ok(CollabOrigin::Client(CollabClient { uid, device_id }))
+        },
+        _ => Err(internal(format!(
+          "couldn't parse collab origin from `{}`",
+          other
+        ))),
+      }
+    },
+  }
+}
+
+#[repr(transparent)]
+#[derive(Copy, Clone, Eq, PartialEq, Default)]
+pub struct UpdateFlags(u8);
+
+impl UpdateFlags {
+  /// Flag bit to mark if update is encoded using [EncoderV2] (if set) or [EncoderV1] (if clear).
+  pub const IS_V2_ENCODED: u8 = 0b0000_0001;
```

Comment:
how about adding the v1_encode:

```rust
  pub const IS_V2_ENCODED: u8 = 0b0000_0001;
  pub const IS_V1_ENCODED: u8 = 0b0000_0100;
  pub const IS_COMPRESSED: u8 = 0b0000_0010;
```

Reply:
These are bit flags. If `IS_V2_ENCODED` is not set, it implies V1 encoding.